syntax = "proto3";

package kvstore;

// The result of a kv-store mutation
message MutationResponse {
    int32 raft_rc = 1;
    string raft_msg = 2;
    optional bytes kvstore_result = 3;
}

// The result of a read to the kv-store
message ReadResponse {
    optional bytes value = 1;
    optional bytes kvstore_result = 2;
}

message Empty {}

message Key {
    bytes key = 1;
}

message KVPair {
    bytes key = 1;
    bytes value = 2;
}

message ServerID {
    int32 server_id = 1;
}

message ClientFacingEndpoint {
    int32 server_id = 1;
    string client_endpoint = 2;
}

message ClusterEndpoints {
    repeated ClientFacingEndpoint endpoints = 1;
}

message ServerEndpoint {
    int32 server_id = 1;
    string raft_endpoint = 2;
    string client_endpoint = 3;
}

service ReplicatedKVStore {
    rpc Get (Key) returns (ReadResponse) {}

    rpc Put (KVPair) returns (MutationResponse) {}

    rpc Update (KVPair) returns (MutationResponse) {}

    rpc Delete (Key) returns (MutationResponse) {}

    rpc GetLeaderID (Empty) returns (ServerID) {}

    rpc GetClusterEndpoints (Empty) returns (ClusterEndpoints) {}
}

// The NuRaft data describing the result of replication
message JoinClusterResult {
    int32 rc = 1;
    string msg = 2;
}

service RaftCluster {
    rpc JoinReplicaSet (ServerEndpoint) returns (JoinClusterResult) {}
}